notebooks/15_flightplanning_tuto.ipynb

```diff
@@ -0,0 +1,1018 @@
+    "        return Value(cost=cost)\n",
+    "\n",
+    "    def _get_initial_state_(self) -> D.T_state:\n",
+    "        \"\"\"\n",
+    "        Get the initial state.\n",
+    "\n",
+    "        Set the start position as initial state.\n",
+    "        \"\"\"\n",
+    "        return self.start\n",
+    "\n",
+    "    def _get_goals_(self) -> Space[D.T_observation]:\n",
+    "        \"\"\"\n",
+    "        Get the domain goals space (finite or infinite set).\n",
+    "\n",
+    "        Set the end position as goal.\n",
+    "        \"\"\"\n",
+    "        return ListSpace([State(None, (self.np - 1, j)) for j in range(self.nc)])\n",
+    "\n",
+    "    def _is_terminal(self, state: State) -> D.T_predicate:\n",
+    "        \"\"\"\n",
+    "        Indicate whether a state is terminal.\n",
+    "\n",
+    "        Stop an episode only when goal reached.\n",
+    "        \"\"\"\n",
+    "        return state.pos[0] == self.np - 1\n",
+    "\n",
+    "    def _get_applicable_actions_from(self, memory: D.T_state) -> Space[D.T_event]:\n",
+    "        \"\"\"\n",
+    "        Get the applicable actions from a state.\n",
+    "        \"\"\"\n",
+    "        x, y = memory.pos\n",
+    "\n",
+    "        space = []\n",
+    "        if x < self.np - 1:\n",
+    "            space.append(Action.straight)\n",
+    "            if y + 1 < self.nc:\n",
+    "                space.append(Action.up)\n",
+    "            if y > 0:\n",
+    "                space.append(Action.down)\n",
+    "\n",
+    "        return ListSpace(space)\n",
+    "\n",
+    "    def _get_action_space_(self) -> Space[D.T_event]:\n",
+    "        \"\"\"\n",
+    "        Define action space.\n",
+    "        \"\"\"\n",
+    "        # return ImplicitSpace(lambda x: isinstance(x, State))\n",
+    "        return EnumSpace(Action)\n",
+    "\n",
+    "    def _get_observation_space_(self) -> Space[D.T_observation]:\n",
+    "        \"\"\"\n",
+    "        Define observation space.\n",
+    "        \"\"\"\n",
+    "        return MultiDiscreteSpace([self.np, self.nc])\n",
+    "\n",
+    "    def _render_from(self, memory: State, **kwargs: Any) -> Any:\n",
+    "        \"\"\"\n",
+    "        Render visually the map.\n",
+    "\n",
+    "        Returns:\n",
+    "            matplotlib figure\n",
+    "        \"\"\"\n",
+    "        # store used matplotlib subplot and image to only update them afterwards\n",
+    "        fig = Figure(figsize=(600, 600))\n",
+    "        fig.canvas.header_visible = False\n",
+    "        fig.canvas.footer_visible = False\n",
+    "        fig.canvas.resizable = False\n",
+    "        fig.set_dpi(1)\n",
+    "\n",
+    "        latmin, latmax = min(self.lat1, self.lat2), max(self.lat1, self.lat2)\n",
+    "        lonmin, lonmax = min(self.lon1, self.lon2), max(self.lon1, self.lon2)\n",
+    "\n",
+    "        ax = plt.axes(\n",
+    "            projection=ccrs.TransverseMercator(\n",
```

Comment:
scatter plot with TransverseMercator projection fails with cartopy 0.18, so maybe we should mention that cartopy0.21 for example is needed